File: 1to150/125_validPalindrom.java
/*
Example 1:

Input: "A man, a plan, a canal: Panama"
Output: true

two pointer扫一次首尾，遇到非字母的跳过，Character.isDigitOrLetter()
遇到字母的首先lower case他们 Character.toLowerCase()
然后比对首位字母是否相同，如果不同则return false
不然在最后return true
*/

class Solution {
    public boolean isPalindrome(String s) {
        if (s == null || s.length() == 0){
            return true;
        }
        
        int start = 0;
        int end = s.length() - 1;
        
        while(start < end){ 
            //首先判断是否是letter如果不是则一直改正           
            while(start < end && !Character.isLetterOrDigit(s.charAt(start))){
                start++;
            }
            //同理对end character做同样的处理
            while(start < end && !Character.isLetterOrDigit(s.charAt(end))){
                end --;
            }
            //上述修改结束后，记录此刻的startCharacter 和 endCharacter
            char cStart = s.charAt(start);
            char cEnd = s.charAt(end);
            
            if (Character.toLowerCase(cStart) != Character.toLowerCase(cEnd)){
                return false;
            }
            start++;
            end--;
        }
        return true;
    }
}

File: 1to150/125_validPalindrom.py
'''
思路依旧还是two pointer
但需要判断此刻的char是不是 alpha or number 不然需要修改index


Example 1:

Input: "A man, a plan, a canal: Panama"
Output: true
'''

class Solution(object):
    def isPalindrome(self, s):
        """
        :type s: str
        :rtype: bool
        """
        if not s:
            return True
        
        start = 0
        end = len(s) - 1
        
        while(start < end):
            while(start < end and not s[start].isalnum()):
                start +=1
            
            while(start < end and not s[end].isalnum()):
                end -= 1
            
            #需要全部小写，不然会出错
            headchar = s[start].lower()
            endchar = s[end].lower()
            # print(headchar, endchar)
            if headchar != endchar:
                return False
            
            start +=1
            end -=1
            
        return True